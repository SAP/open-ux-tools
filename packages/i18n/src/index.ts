export { getCapI18nBundle, getPropertiesI18nBundle } from './read';
export { createCapI18nEntries, createPropertiesI18nEntries } from './write';

export {
    getI18nFolderNames,
    getI18nMaxLength,
    getI18nTextType,
    extractI18nKey,
    extractDoubleCurlyBracketsKey,
    getI18nUniqueKey,
    convertToCamelCase,
    convertToPascalCase,
    printPropertiesI18nEntry,
    printPropertiesI18nAnnotation
} from './utils';

export {
    CdsEnvironment,
    CdsI18nConfiguration,
    CdsI18nEnv,
    I18nAnnotation,
    I18nAnnotationNode,
    I18nBundle,
    I18nEntry,
    NewI18nEntry,
    SapLongTextType,
    SapShortTextType,
    SapTextType,
    ValueNode
} from './types';

export { csvToI18nBundle, jsonToI18nBundle, propertiesToI18nEntry } from './transformer';
