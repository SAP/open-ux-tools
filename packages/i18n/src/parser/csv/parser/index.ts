export { parseCsv } from './csv-parser';
