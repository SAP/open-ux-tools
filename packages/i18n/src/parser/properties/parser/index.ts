import { tokenize } from '../lexer';
import type { PropertiesParseResult } from '../types';
import { getPropertyList } from './properties-parser';

export const parseProperties = (text: string): PropertiesParseResult => {
    const tokens = tokenize(text);
    const ast = getPropertyList(tokens, text);
    return { ast, tokens };
};
