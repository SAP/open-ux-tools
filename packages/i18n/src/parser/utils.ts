import { Range, Position } from 'vscode-languageserver-types';

export { Range, Position };

/**
 * Computes line offsets for the given string.
 *
 * @param text text
 * @returns array of number
 */
export function getLineOffsets(text: string): number[] {
    const lineOffsets: number[] = [0];
    for (let index = 0; index < text.length; ) {
        const character = text[index];
        if (character === '\n') {
            lineOffsets.push(index + 1);
            index++; // Increment index here
        } else if (character === '\r') {
            if (index + 1 < text.length && text[index + 1] === '\n') {
                index += 2; // Increment index by 2 when encountering '\r\n'
            } else {
                index++; // Increment index by 1 when encountering '\r'
            }
            lineOffsets.push(index); // Push the updated index value
        } else {
            index++; // Increment index if character is not a line break
        }
    }
    return lineOffsets;
}

/**
 * Position at.
 *
 * @param lineOffsets Array of indices with line start offsets.
 * e.g [0] represents a document with one line that starts at offset 0.
 * @param offset offset
 * @param textLength max length
 * @returns position
 */
export function positionAt(lineOffsets: number[], offset: number, textLength: number): Position {
    const target = Math.max(Math.min(offset, textLength), 0);
    let low = 0;
    let high = lineOffsets.length;

    if (high === 0) {
        return Position.create(0, target);
    }

    while (low < high) {
        const mid = Math.floor((low + high) / 2);
        if (lineOffsets[mid] > target) {
            high = mid;
        } else {
            low = mid + 1;
        }
    }
    const line = low - 1;
    return Position.create(line, target - lineOffsets[line]);
}

/**
 * Create range.
 *
 * @param lineOffsets line offset
 * @param start start point
 * @param end end point
 * @param textLength text length
 * @returns range
 */
export function rangeAt(lineOffsets: number[], start: number, end: number, textLength: number): Range {
    return Range.create(positionAt(lineOffsets, start, textLength), positionAt(lineOffsets, end, textLength));
}
