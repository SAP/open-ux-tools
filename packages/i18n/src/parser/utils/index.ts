export { getLineOffsets } from './line-offsets';
export { positionAt } from './position';
export { rangeAt } from './range';
export { Position, Range } from './language-server';
