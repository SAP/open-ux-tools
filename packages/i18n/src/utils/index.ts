export { getI18nConfiguration, getI18nFolderNames } from './config';
export { getCapI18nFiles, resolveCapI18nFolderForFile, getCapI18nFolder } from './resolve';
export { csvPath, jsonPath, capPropertiesPath, doesExist } from './path';
export { printPropertiesI18nAnnotation, printPropertiesI18nEntry } from './print';
export {
    getI18nMaxLength,
    getI18nTextType,
    applyIndent,
    discoverIndent,
    discoverLineEnding,
    convertToCamelCase,
    convertToPascalCase
} from './text';
export { extractI18nKey, getI18nUniqueKey, extractDoubleCurlyBracketsKey } from './key';
export { readFile, writeFile } from './mem-fs-editor';
