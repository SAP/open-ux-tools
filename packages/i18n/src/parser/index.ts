import { FileFormat, ParseResult } from './types';
import { parseProperties } from './properties/parser';
import { parseCsv } from './csv/parser';

export const parse = (text: string, format: FileFormat): ParseResult => {
    if (format === FileFormat.properties) {
        return parseProperties(text);
    }
    return parseCsv(text);
};
