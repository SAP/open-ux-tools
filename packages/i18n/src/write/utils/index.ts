import type { NewI18nEntry } from '../../types';
import { printPropertiesI18nEntry, readFile, writeFile } from '../../utils';
import type { Editor } from 'mem-fs-editor';

/**
 * Write i18n entries to an existing i18n.properties file.
 *
 * @param i18nFilePath i18n file path
 * @param newI18nEntries  new i18n entries that will be maintained
 * @param fs optional `mem-fs-editor` instance. If provided, `mem-fs-editor` api is used instead of `fs` of node
 * @returns boolean
 */
export async function writeToExistingI18nPropertiesFile(
    i18nFilePath: string,
    newI18nEntries: NewI18nEntry[],
    fs?: Editor
): Promise<boolean> {
    let newContent = newI18nEntries
        .map((entry) => printPropertiesI18nEntry(entry.key, entry.value, entry.annotation))
        .join('');

    const content = await readFile(i18nFilePath, fs);
    const lines = content.split(/\r\n|\n/);
    // check if file does not end with new line
    if (lines.length > 0 && lines[lines.length - 1].trim()) {
        // If there no end line - add new gap line before new content
        newContent = `\n${newContent}`;
    }
    await writeFile(i18nFilePath, content.concat(newContent), fs);
    return true;
}
