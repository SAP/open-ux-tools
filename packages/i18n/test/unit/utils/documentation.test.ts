import { type I18nEntry, SapShortTextType, initI18n, getI18nDocumentation } from '../../../src';
import { Range } from '../../../src/parser/utils';

describe('documentation', () => {
    describe('getI18nDocumentation', () => {
        const range = Range.create(0, 0, 0, 0);
        beforeAll(async () => {
            await initI18n();
        });
        test('full annotations', () => {
            // arrange
            const input: I18nEntry = {
                filePath: 'absolute/file/path',
                key: {
                    value: 'productDetailsInfo',
                    range
                },
                value: {
                    value: 'product details info',
                    range
                },
                annotation: {
                    textType: {
                        value: SapShortTextType.Label,
                        range
                    },
                    maxLength: {
                        value: 60,
                        range
                    },
                    note: {
                        value: ' Label for a section',
                        range
                    }
                }
            };
            // act
            const result = getI18nDocumentation(input);
            // assert
            expect(result).toMatchSnapshot();
        });
        test('no note', () => {
            // arrange
            const input: I18nEntry = {
                filePath: 'absolute/file/path',
                key: {
                    value: 'productDetailsInfo',
                    range
                },
                value: {
                    value: 'product details info',
                    range
                },
                annotation: {
                    textType: {
                        value: SapShortTextType.Label,
                        range
                    },
                    maxLength: {
                        value: 60,
                        range
                    }
                }
            };
            // act
            const result = getI18nDocumentation(input);
            // assert
            expect(result).toMatchSnapshot();
        });
        test('no max length', () => {
            // arrange
            const input: I18nEntry = {
                filePath: 'absolute/file/path',
                key: {
                    value: 'productDetailsInfo',
                    range
                },
                value: {
                    value: 'product details info',
                    range
                },
                annotation: {
                    textType: {
                        value: SapShortTextType.Label,
                        range
                    }
                }
            };
            // act
            const result = getI18nDocumentation(input);
            // assert
            expect(result).toMatchSnapshot();
        });
        test('no annotation', () => {
            // arrange
            const input: I18nEntry = {
                filePath: 'absolute/file/path',
                key: {
                    value: 'productDetailsInfo',
                    range
                },
                value: {
                    value: 'product details info',
                    range
                }
            };
            // act
            const result = getI18nDocumentation(input);
            // assert
            expect(result).toMatchSnapshot();
        });
    });
});
