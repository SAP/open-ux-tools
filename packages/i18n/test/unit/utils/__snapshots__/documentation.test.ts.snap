// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentation getI18nDocumentation full annotations 1`] = `
"**Text Key:** productDetailsInfo

**Text Value:** product details info

**Additional Information:** XFLD,60: Label for a section"
`;

exports[`documentation getI18nDocumentation no annotation 1`] = `
"**Text Key:** productDetailsInfo

**Text Value:** product details info"
`;

exports[`documentation getI18nDocumentation no max length 1`] = `
"**Text Key:** productDetailsInfo

**Text Value:** product details info

**Additional Information:** XFLD"
`;

exports[`documentation getI18nDocumentation no note 1`] = `
"**Text Key:** productDetailsInfo

**Text Value:** product details info

**Additional Information:** XFLD,60"
`;
